brainiak/reprsimil/brsa.py

```diff
@@ -1669,3 +2290,1292 @@ def _loglike_AR1_singpara(self, param, XTX, XTDX, XTFX, YTY_diag,
+        # This function calculates the log likelihood of data given AR(1)
+        # parameters of noise as free parameters.
+        # Free parameters are in param.
+        # It serves as a null model which assumes no response to design
+        # matrix.
+        a1 = param
+        rho1 = 2.0 / np.pi * np.arctan(a1)  # auto-regressive coefficients
+
+        YTAY = self._make_sandwidge(YTY_diag, YTDY_diag, YTFY_diag, rho1)
+        # dimension: space,
+        # A/sigma2 is the inverse of noise covariance matrix in each voxel.
+        # YTAY means Y'AY
+        X0TAX0 = X0TX0[np.newaxis, :, :] - rho1[:, np.newaxis, np.newaxis] \
+            * X0TDX0[np.newaxis, :, :] \
+            + rho1[:, np.newaxis, np.newaxis]**2 * X0TFX0[np.newaxis, :, :]
+        # dimension: space*#baseline*#baseline
+        X0TAY = self._make_sandwidge(X0TY, X0TDY, X0TFY, rho1)
+        # dimension: #baseline*space
+        # X0TAX0_i = np.linalg.solve(X0TAX0, np.identity(n_X0)[None, :, :])
+        X0TAX0_i = np.linalg.inv(X0TAX0)
+        # dimension: space*#baseline*#baseline
+        YTAcorrY = YTAY - np.sum(X0TAY * np.einsum('ijk,ki->ji',
+                                                   X0TAX0_i, X0TAY), axis=0)
+        # dimension: space,
+        sigma2 = YTAcorrY / (n_T - n_X0)
+        # dimension: space,
+        LL = - np.sum(np.log(sigma2)) * (n_T - n_X0) * 0.5 \
+            + np.sum(np.log(1 - rho1**2)) * n_run * 0.5 \
+            - np.sum(np.log(np.linalg.det(X0TAX0))) * 0.5 \
+            - (n_T - n_X0) * n_V * (1 + np.log(2 * np.pi)) * 0.5
+        # The following are for calculating the derivative to a1
+        deriv_a1 = np.empty(n_V)
+        dYTAY_drho1 = self._make_sandwidge_grad(YTDY_diag, YTFY_diag, rho1)
+        # dimension: space,
+        dX0TAX0_drho1 = - X0TDX0 \
+            + 2 * rho1[:, np.newaxis, np.newaxis] * X0TFX0
+        # dimension: space*rank*rank
+        dX0TAY_drho1 = self._make_sandwidge_grad(X0TDY, X0TFY, rho1)
+        # dimension: rank*space
+
+        # The following are executed for each voxel.
+        for i_v in range(n_V):
+            # All variables with _ele as suffix are for data of just one voxel
+            invX0TAX0_X0TAY_ele = np.dot(X0TAX0_i[i_v, :, :], X0TAY[:, i_v])
+            # preparation for the variable below
+            dYTAcorrY_drho1_ele = dYTAY_drho1[i_v] \
+                - np.dot(dX0TAY_drho1[:, i_v], invX0TAX0_X0TAY_ele) * 2\
+                + np.dot(np.dot(invX0TAX0_X0TAY_ele, dX0TAX0_drho1[i_v, :, :]),
+                         invX0TAX0_X0TAY_ele)
+            deriv_a1[i_v] = 2 / np.pi / (1 + a1[i_v]**2) \
+                * (- n_run * rho1[i_v] / (1 - rho1[i_v]**2)
+                   - np.einsum('ij,ij', X0TAX0_i[i_v, :, :],
+                               dX0TAX0_drho1[i_v, :, :]) * 0.5
+                   - dYTAcorrY_drho1_ele * 0.5 / sigma2[i_v])
+        deriv = deriv_a1
+
+        return -LL, -deriv
+
+
+class GBRSA(BRSA):
```

Comment:
Hi @mihaic I realized that there is still some difference between the two docstring. So I only added the reference without removing the basic description of GBRSA.